pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java

```diff
@@ -110,94 +110,99 @@ public void init(HelixManager helixManager) {
   }
 
   @Override
-  public synchronized void processClusterChange(HelixConstants.ChangeType changeType) {
-    Preconditions.checkState(changeType == HelixConstants.ChangeType.EXTERNAL_VIEW
-        || changeType == HelixConstants.ChangeType.INSTANCE_CONFIG, "Illegal change type: %s", changeType);
-    if (changeType == HelixConstants.ChangeType.EXTERNAL_VIEW) {
-      processExternalViewChange();
-    } else {
+  public synchronized void processClusterChange(ChangeType changeType) {
+    if (changeType == ChangeType.IDEAL_STATE || changeType == ChangeType.EXTERNAL_VIEW) {
```

Comment:
Does that mean broker will do segment assignment two times, one on IS change and the other one on EV change?

Reply:
Segment assignment is for the rebalancing, not for the routing callback updates. Routing updates are much cheaper. Broker will update the mapping from servers to segments on both IS and EV changes. IS change is less frequent than EV change, and we cache the version of them so we won't redo the work if the change is already processed, so I don't worry too much about the overhead.